flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala

```diff
@@ -411,6 +411,23 @@ class TaskManager(
             log.debug(s"Cannot find task to fail for execution ${executionID})")
           }
 
+        // stops a task
+        case StopTask(executionID) =>
+          val task = runningTasks.get(executionID)
+          if (task != null) {
+            try {
+              task.stopExecution()
```

Comment:
I think that's already how sources are supposed to behave. Kafka does it as well, the only thing that needs to be changed is removing the `consumer.shutdown()` call from `cancel()`.